README.md
# apoclypsebm - The ApoCLypse Bitcoin Miner
## Background
This hobby project maintained by
[Justin T. Arthur](https://github.com/JustinTArthur) undertakes the quixotic
task of maintaining a modern Bitcoin miner for programmable compute devices like
GPUs. It was forked from the PyOpenCL Bitcoin Miner (poclbm), a project authored
by [m0mchil](https://github.com/m0mchil) and contributors.

It features an OpenCL Kernel that has incorporated ideas or code from:
* [diapolo](https://github.com/diapolo)
* [m0mchil](https://github.com/m0mchil)
* [neurobox](https://bitcointalk.org/index.php?action=profile;u=106397)
* [phataeus](https://sourceforge.net/u/phateus/)
* [rethaw](https://bitcointalk.org/index.php?action=profile;u=18618)

If your work is represented herein and I didn't give you credit, please let me
know. At the moment, I reserve no rights to the mining driver or the OpenCL
kernel. They were derived from public domain works.

## Economy
At the time of writing, on-chip implementations of the Bitcoin mining algorithm
will outperform this software in both time and joules expended. Under most
conditions, mining blocks on a Bitcoin chain where on-chip implementations are
competing would be at a tremendous waste of expended resources.

## Maintenance Notes
It looks like the work sourcing threads run into i/o issues occasionally due to
using the not-thread-safe Python http lib. I don't aim to address this as most
of the threaded communication ought to be completely replaced by an event runner
like asyncio or trio at some point.

Thanks to @momchil for the original `getwork` code, @luke-jr @sipa and @vsergeev
for helping me understand getblocktemplate. 

## Installation
In an environment with Python 3.5+:

    pip3 install apoclypsebm

## Usage
    apoclypse [OPTION]... SERVER[#tag]...

`SERVER` is one or more [http[s]|stratum://]user:pass@host:port (required)  
[#tag] is an optional per server user-friendly name displayed in stats.

### Options
```
Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  --verbose             verbose output, suitable for redirection to log file
  -q, --quiet           suppress all output except hash rate display
  --proxy=PROXY         specify as
                        [[socks4|socks5|http://]user:pass@]host:port (default
                        proto is socks5)
  --no-ocl              don't use OpenCL
  --no-bfl              don't use Butterfly Labs
  --stratum-proxies     search for and use stratum proxies in subnet
  -d DEVICE, --device=DEVICE
                        comma separated device IDs, by default will use all
                        (for OpenCL - only GPU devices)
  -a ADDRESS, --address=ADDRESS
                        Bitcoin address to spend the block reward to if
                        allowed. Required for solo mining, ignored with
                        stratum or getwork sources.
  --coinbase-msg=COINBASE_MSG
                        Custom text to include in the coinbase of the
                        generation tx if allowed, encoded as UTF-8.
                        default=ApoCLypse

  Miner Options:
    -r RATE, --rate=RATE
                        hash rate display interval in seconds, default=1 (60
                        with --verbose)
    -e ESTIMATE, --estimate=ESTIMATE
                        estimated rate time window in seconds, default 900 (15
                        minutes)
    -t TOLERANCE, --tolerance=TOLERANCE
                        use fallback pool only after N consecutive connection
                        errors, default 2
    -b FAILBACK, --failback=FAILBACK
                        attempt to fail back to the primary pool after N
                        seconds, default 60
    --cutoff-temp=CUTOFF_TEMP
                        AMD GPUs, BFL only. For GPUs requires
                        github.com/mjmvisser/adl3. Comma separated
                        temperatures at which to skip kernel execution, in C,
                        default=95
    --cutoff-interval=CUTOFF_INTERVAL
                        how long to not execute calculations if CUTOFF_TEMP is
                        reached, in seconds, default=0.01
    --no-server-failbacks
                        disable using failback hosts provided by server

  OpenCL Options:
    Every option except 'platform' and 'vectors' can be specified as a
    comma separated list. If there aren't enough entries specified, the
    last available is used. Use --vv to specify per-device vectors usage.

    -p PLATFORM, --platform=PLATFORM
                        use platform by id
    -k KERNEL, --kernel=KERNEL
                        OpenCL Kernel to use. Defaults to apoclypse-0
    -w WORKSIZE, --worksize=WORKSIZE
                        work group size, default is maximum reported by the
                        driver.
    -f FRAMES, --frames=FRAMES
                        will try to bring single kernel execution to 1/frames
                        seconds, default=30, increase this for less desktop
                        lag
    -s FRAME_SLEEP, --sleep=FRAME_SLEEP
                        sleep per frame in seconds, default 0
    --vv=VECTORS        Specifies size of SIMD vectors per selected device.
                        Only size 0 (no vectors) and 2 supported for now.
                        Comma separated for each device. e.g. 0,2,2
    -v, --vectors       Use 2-item vectors for all devices.
```

### Examples
Solo mining against a Bitcoin Core node's RPC port:

    apoclypse --address bc1qf2277gpv3hlewlqq2cuvf77qz5xcjzr7njf3s9 --verbose http://u:p@127.0.0.1:8332

Mining on OpenCL platform 0, device 1 against a stratum server:

    apoclypse -p 0 -d 1 --verbose stratum://u:p@us-east.stratum.hushpool.io:3333


CHANGES.md
## New in Version 1.1.4
* Added `-k`/`--kernel` option for specifying which of available kernels to
use. Only `apoclypse-0` and `apoclypse-loopy` are available at the moment.
* New "apoclypse-loopy" kernel is an alternative implementation of the
poclbm/phoenix/phatk kernel that uses for-loops. On some older platforms, this
provides better performance. On modern AMD and nVidia software+hardware
stacks, the basic apoclypse-0 kernel is fine.
* The `-w`/`--worksize` option now uses kernel code to configure the work
group size when the OpenCL stack tries to enforce a smaller maximum size. This
allows for up to 1024-sized work groups on recent AMD and nVidia
drivers+devices.
* Fix for discrete cards doing wasted work because output buffer wasn't being
copied back to host correctly.
* Fix for when stratum servers supply a floating point difficulty number
[#7](JustinTArthur/apoclypsebm#7). In this case, we floor the result of any
arithmetic performed against it.
* Use monotonic timers for rate calculations so that system time changes
don't impact them.
* Upgrade to latest PyOpenCL now that pybind11 is fixing the header
situation.

## New in Version 1.1.3
Bug fix release. Installation from PyPI (e.g. with pip) had some
glaring issues. Additionally, PyOpenCL removed deprecated APIs we were using.
I've pinned us to PyOpenCL releases prior to their pybind11 integration so that
this project remains installable from pip (see
[inducer/pyopencl#278](https://github.com/inducer/pyopencl/issues/278)).

## New in Version 1.1.0
This release focused on supporting the `getblocktemplate` call proposed in
BIP 22 and BIP 23 and implemented by Bitcoin Core and btcd. This finally allows
mining without a pool if you supply an address to mine coins to at the command
line.

HTTP work sources now default to using `getblocktemplate` instead of `getwork`.
This feature is new and might have bugs that could lead to loss of potential
mining rewards.


apoclypsebm.py
#!/usr/bin/env python

from Switch import Switch
from optparse import OptionGroup, OptionParser
from time import sleep
from util import tokenize
from version import VERSION
import log
import socket


class LongPollingSocket(socket.socket):
    """
    Socket wrapper to enable socket.TCP_NODELAY and KEEPALIVE
    """

    def __init__(self, family=socket.AF_INET, type=socket.SOCK_STREAM, proto=0):
        super(LongPollingSocket, self).__init__(family, type, proto)
        if type == socket.SOCK_STREAM:
            self.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            self.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
        self.settimeout(5)

# Monkey patch our wrapped socket:
socket.socket = LongPollingSocket

usage = "usage: %prog [OPTION]... SERVER[#tag]...\nSERVER is one or more [http[s]|stratum://]user:pass@host:port " \
        "         (required)\n[#tag] is a per-server user friendly name displayed in stats (optional)"
parser = OptionParser(version=VERSION, usage=usage)
parser.add_option('--verbose', dest='verbose', action='store_true',
                  help='verbose output, suitable for redirection to log file')
parser.add_option('-q', '--quiet', dest='quiet', action='store_true',
                  help='suppress all output except hash rate display')
parser.add_option('--proxy', dest='proxy', default='',
                  help='specify as [[socks4|socks5|http://]user:pass@]host:port (default proto is socks5)')
parser.add_option('--no-ocl', dest='no_ocl', action='store_true', help="don't use OpenCL")
parser.add_option('--no-bfl', dest='no_bfl', action='store_true', help="don't use Butterfly Labs")
parser.add_option('--stratum-proxies', dest='stratum_proxies', action='store_true',
                  help="search for and use stratum proxies in subnet")
parser.add_option('-d', '--device', dest='device', default=[],
                  help='comma separated device IDs, by default will use all (for OpenCL - only GPU devices)')

group = OptionGroup(parser, "Miner Options")
group.add_option('-r', '--rate', dest='rate', default=1,
                 help='hash rate display interval in seconds, default=1 (60 with --verbose)', type='float')
group.add_option('-e', '--estimate', dest='estimate', default=900,
                 help='estimated rate time window in seconds, default 900 (15 minutes)', type='int')
group.add_option('-t', '--tolerance', dest='tolerance', default=2,
                 help='use fallback pool only after N consecutive connection errors, default 2', type='int')
group.add_option('-b', '--failback', dest='failback', default=60,
                 help='attempt to fail back to the primary pool after N seconds, default 60', type='int')
group.add_option('--cutoff-temp', dest='cutoff_temp', default=[],
                 help='AMD GPUs, BFL only. For GPUs requires github.com/mjmvisser/adl3. Comma separated temperatures at'
                      ' which to skip kernel execution, in C, default=95')
group.add_option('--cutoff-interval', dest='cutoff_interval', default=[],
                 help='how long to not execute calculations if CUTOFF_TEMP is reached, in seconds, default=0.01')
group.add_option('--no-server-failbacks', dest='nsf', action='store_true',
                 help='disable using failback hosts provided by server')
parser.add_option_group(group)

group = OptionGroup(parser,
                    "OpenCL Options",
                    "Every option except 'platform' and 'vectors' can be specified as a comma separated list. "
                    "If there aren't enough entries specified, the last available is used. "
                    "Use --vv to specify per-device vectors usage."
                    )
group.add_option('-p', '--platform', dest='platform', default=-1, help='use platform by id', type='int')
group.add_option('-w', '--worksize', dest='worksize', default=[],
                 help='work group size, default is maximum returned by OpenCL')
group.add_option('-f', '--frames', dest='frames', default=[],
                 help='will try to bring single kernel execution to 1/frames seconds, default=30, increase this for'
                      ' less desktop lag')
group.add_option('-s', '--sleep', dest='frameSleep', default=[], help='sleep per frame in seconds, default 0')
group.add_option('--vv', dest='vectors', default=[], help='use vectors, default false')
group.add_option('-v', '--vectors', dest='old_vectors', action='store_true', help='use vectors')
parser.add_option_group(group)


def main():
    options, options.servers = parser.parse_args()

    log.verbose = options.verbose
    log.quiet = options.quiet

    options.rate = max(options.rate, 60) if options.verbose else max(options.rate, 0.1)

    options.version = VERSION

    options.max_update_time = 60

    options.device = tokenize(options.device, 'device', [])

    options.cutoff_temp = tokenize(options.cutoff_temp, 'cutoff_temp', [95], float)
    options.cutoff_interval = tokenize(options.cutoff_interval, 'cutoff_interval', [0.01], float)

    switch = None
    try:
        switch = Switch(options)

        if not options.no_ocl:
            import OpenCLMiner

            for miner in OpenCLMiner.initialize(options):
                switch.add_miner(miner)

        if not options.no_bfl:
            import BFLMiner

            for miner in BFLMiner.initialize(options):
                switch.add_miner(miner)

        if not switch.servers:
            print('\nAt least one server is required\n')
        elif not switch.miners:
            print('\nNothing to mine on, exiting\n')
        else:
            for miner in switch.miners:
                miner.start()
            switch.loop()
    except KeyboardInterrupt:
        print('\nbye')
    finally:
        for miner in switch.miners:
            miner.stop()
        if switch:
            switch.stop()

        if not options.no_ocl:
            OpenCLMiner.shutdown()
    sleep(1.1)

if __name__ == "__main__":
    main()